### DISCUSSION TITLE
WebQuery to get the csv string, parse it into string segments and store as Var1

### DISCUSSION THREAD

**Discussion by jgrulich** [arendst/Tasmota#22418] [Q&A] [answered]
Hello, I' trying to write the simple Thermostat rule which will download the temperature from the csv txt string.
The rule is:
Rule1
  ON WebQuery http://server.lan GET DO Var1 %value% ENDON
  ON Var1#State DO Var2=(20.0-%Var1%)*10.0+30.0 ENDON
  ON Var2#State > 45.0 DO Var2=45.0 ENDON
  ON Var2#State < 30.0 DO Var2=30.0 ENDON
  ON Var3#State DO ot_tboiler %Var2% ENDON

But the response of the WebQuery is only RSL: RESULT = {"WebQuery":"Done"}
No data from the csv string.
When I enter http://server.lan/csv into the web browser, I receive text string: 21.7;49.1;1033.7;0;0;0
I need to get the first value "21.7", store it as Var1 and do the rest in the rule and finally set the new temperature.
I'm totally new in the Rules writing and Tasmota itself and I've tried whatever I was able, but no lucky.
Please be so kind and help to solve this for me.

**Comment by sfromis** [comment]
First, you cannot "embed" commands in the trigger part of a rule. All commands needs to be issued separately, and then you can have rule triggers pick up JSON data resulting from the command. And no, that does not mean that the server response needs to be JSON. If you issue the `WebQuery` command from a rule, it needs to be prefixed as `Backlog WebQuery http.....`

To even have the `WebQuery` response instead of just "Done", you need a build with `USE_WEBSEND_RESPONSE`

However, with rules you have no features suitable for CSV parsing (or any other string parsing). It would be workable if you have a server endpoint only returning one value.

With ESP32, I can recommend the Berry language providing good features for string parsing. I think that it might also be doable with the scripter feature replacing rules on ESP8266, but that's a very different beast.

**Comment by jgrulich** [comment]
I've rebuilt with USE_WEBSEND_RESPONSE and modified the csv response, but still the Var1 is not set to the WebQuery response:
CMD: backlog WebQuery http://server.lan/csv GET; Var1 %value%
RSL: RESULT = {"WebQuery":"22.9"}
RSL: RESULT = {"Var1":"%value%"}

I've tried different ways for it:

CMD: backlog WebQuery http://server.lan/csv GET; Var1 WebQuery; Var1
12:08:28.944 RSL: RESULT = {"WebQuery":"22.9"}
12:08:28.987 RSL: RESULT = {"Var1":"WebQuery"}
12:08:29.237 RSL: RESULT = {"Var1":"WebQuery"}

**Reply by sfromis** [reply]
As I told you, you can have a **separate** rule to pick data from the JSON showing the `WebQuery` result (as you now have). You cannot do anything with `%value%` on the command line, only as part of a rule, referring to the trigger, like `WebQuery#data`

What I told you about needing to use `Backlog` was that you need this when you issue a command  from a rule, and want rule triggers to act on the response, not to have any more commands at the same time as `WebQuery`

**Comment by jgrulich** [comment]
I've modified it as follows and it seems to work, except the ot_tboiler command is executed twice.

`14:05:02.548 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":294,"Free":217,"Rules":"ON Time#Minute DO backlog WebQuery http://opinas.lan/airswitch/temp.json GET ENDON ON WebQuery#Data DO Var1 %value% ENDON ON Var1#State DO Var2=(20.0-%Var1%)*10.0+30.0 ENDON ON Var2#State > 45.0 DO Var2=45.0 ENDON ON Var2#State < 30.0 DO Var2=30.0 ENDON ON Var2#State DO ot_tboiler %Var2% ENDON"}}`

14:08:00.285 RUL: TIME#MINUTE performs 'backlog WebQuery http://opinas.lan/airswitch/temp.json GET'
14:08:00.336 RSL: RESULT = {"WebQuery":"22.6"}
14:08:00.352 RUL: WEBQUERY#DATA performs 'Var1 22.6'
14:08:00.357 RSL: RESULT = {"Var1":"22.6"}
14:08:00.382 RUL: VAR1#STATE performs 'Var2=(20.0-22.6)*10.0+30.0'
14:08:00.388 RSL: RESULT = {"Var2":"4.000"}
14:08:00.418 RUL: VAR2#STATE < 30.0 performs 'Var2=30.0'
14:08:00.423 RSL: RESULT = {"Var2":"30.000"}
14:08:00.438 RUL: VAR2#STATE performs 'ot_tboiler 30.000'
14:08:00.443 RSL: RESULT = {"ot_tboiler":30.0}
14:08:00.503 RUL: VAR2#STATE performs 'ot_tboiler 30.000'
14:08:00.509 RSL: RESULT = {"ot_tboiler":30.0}

**Reply by sfromis** [reply]
Each time you modify `var2`, the trigger will fire.

**Reply by barbudor** [reply]
You can manage that using IF conditions on var1 to directly set Var2 to the right value

**Comment by jgrulich** [comment, marked as answer]
Finally, I make it working. Thank @sfromis to be patient with me and you help and effort.

This is the final rule:

`Rule1
ON Time#Minute DO backlog WebQuery http://opinas.lan/airswitch/temp.json GET ENDON
ON WebQuery#Data DO Var1 %value% ENDON
ON Var1#State DO Var2=(%Mem1%-%Var1%)*10.0+30.0 ENDON
ON Var2#State < 30.0 DO ot_ch 0 ENDON
ON Var2#State >= 30.0 DO ot_ch 1 ENDON
ON Var2#State > 45.0 DO Var2=45.0 ENDON
ON Var2#State >= 30.0 DO ot_tboiler %Var2% ENDON`

It switches OFF heating when the temp request is below 30' and on if above and set the requested temp then.

**Reply by sfromis** [reply]
You will still get the `ot_tboiler` command executed twice if the result of the `var2` calculation is > 45.0, as you then set `var2` again.
